src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt

```diff
@@ -243,82 +219,192 @@ class IncomingPaymentHandler(val nodeParams: NodeParams, val db: IncomingPayment
+                                                    willAddHtlcParts.forEach { add(IncomingPayment.ReceivedWith.AddedToFeeCredit(it.amount)) }
+                                                }
+                                                val actions = buildList {
+                                                    // We send a single add_fee_credit for the will_add_htlc set.
+                                                    add(SendOnTheFlyFundingMessage(AddFeeCredit(nodeParams.chainHash, incomingPayment.preimage)))
+                                                    htlcParts.forEach { add(WrappedChannelCommand(it.htlc.channelId, ChannelCommand.Htlc.Settlement.Fulfill(it.htlc.id, incomingPayment.preimage, true))) }
+                                                }
+                                                acceptPayment(incomingPayment, receivedWith, actions)
+                                            }
+                                            else -> {
+                                                // We're not adding to our fee credit, so we need to check our liquidity policy.
+                                                // Even if we have enough fee credit to pay the fees, we may want to wait for a lower feerate.
+                                                when (val rejected = nodeParams.liquidityPolicy.value.maybeReject(
+                                                    requestedAmount.toMilliSatoshi(),
+                                                    fundingRate.fees(currentFeerate, requestedAmount, requestedAmount).total.toMilliSatoshi(),
+                                                    LiquidityEvents.Source.OffChainPayment,
+                                                    logger
+                                                )) {
+                                                    is LiquidityEvents.Rejected -> rejectOnTheFlyFunding(payment, incomingPayment, rejected)
+                                                    else -> {
+                                                        val actions = listOf(OpenOrSplicePayment(payment.amountReceived, requestedAmount, fundingRate, incomingPayment.preimage))
+                                                        val paymentOnlyHtlcs = payment.copy(
+                                                            // We need to splice before receiving the remaining HTLC parts.
+                                                            // We extend the duration of the MPP timeout to give more time for funding to complete.
+                                                            startedAtSeconds = payment.startedAtSeconds + 30,
+                                                            // We keep the currently added HTLCs, and should receive the remaining HTLCs after the open/splice.
+                                                            parts = htlcParts.toSet()
```

Comment:
👓  This is where we clean up the `WillAddHtlcPart`s and wait for the actual htlc parts to arrive.